ssl/quic/quic_rstream.c

```diff
@@ -132,3 +181,91 @@ int ossl_quic_rstream_available(QUIC_RSTREAM *qrs, size_t *avail, int *fin)
 #endif
     return 1;
 }
+
+int ossl_quic_rstream_get_record(QUIC_RSTREAM *qrs,
+                                 const unsigned char **record, size_t *rec_len,
+                                 int *fin)
+{
+    const unsigned char *record_ = NULL;
+    size_t rec_len_, max_len;
+
+    if (!ossl_sframe_list_lock_head(&qrs->fl, &qrs->head_range, &record_, fin)) {
+        *record = NULL;
+        *rec_len = 0;
+        return 0;
+    }
+
+    rec_len_ = (size_t)(qrs->head_range.end - qrs->head_range.start);
+
+    if (record_ == NULL && rec_len_ != 0) {
+        record_ = ring_buf_get_ptr(&qrs->rbuf, qrs->head_range.start,
+                                   &max_len);
+        if (!ossl_assert(record_ != NULL))
+            return 0;
+        if (max_len < rec_len_) {
+            rec_len_ = max_len;
+            qrs->head_range.end = qrs->head_range.start + max_len;
+        }
+    }
+
+    *rec_len = rec_len_;
+    *record = record_;
+    return 1;
+}
+
+
+int ossl_quic_rstream_release_record(QUIC_RSTREAM *qrs, size_t read_len)
+{
+    uint64_t offset;
+    OSSL_TIME rtt = get_rtt(qrs);
+
+    if (!ossl_sframe_list_is_head_locked(&qrs->fl))
+        return 0;
+
+    if (read_len > qrs->head_range.end - qrs->head_range.start)
+        offset = qrs->head_range.end;
```

Comment:
It might be better to just fail here, since this should be a caller error.

Reply:
If you look at the description I thought it might be convenient to call the function with SIZE_MAX to release the whole record.

Reply:
Hmm, okay. Could we handle that separately as an explicit sentinel value in the code? If the caller passes some wrong value which is not `SIZE_MAX` I think it should still fail.

Reply:
OK, good idea.

Reply:
Done